### PR TITLE
Split up kimchi bindings

### PR DESCRIPTION
**Pull request by mrmr1993**
This splits up kimchi bindings into 3 stages:
* `kimchi_types`, the pure OCaml types that the rust functions consume and return;
* `pasta_bindings`, the bindings for field and group elements, for which we have equivalents in snarkette;
* `kimchi_bindings`, the bindings for functions related to the kimchi proof system.

The intention is that this splitting will make it far easier to remove pickles -- and thus the rust backend -- as a dependency for everything up to the blockchain/transaction snark modules. Once this is done, it should be straightforward to create a virtual module system for the
pasta curve and field types, and thus to compile the client SDK directly from `mina_base` and friends, without further need for optional compilation.

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Comment by mrmr1993** [comment]
A potential path forward for easy client SDK compilation on top of this:
* split the curve/field parts of `kimchi_backend` into their own module
* turn that module into a 'dune virtual module' whose alternative implementation is provided by snarkette (similar to the implementation of `snark_params_nonconsensus`)
* make a few small tweaks to snarky to remove the need for `Snark0.Make` to take a constraint system
* remove the dependency on `pickles` from `Snark_params`, `Mina_base`, etc.
  - this will likely require uplifting the `Proof.Repr` type out of the `pickles` library, but that should be fairly straightforward.

File: src/lib/crypto/kimchi_bindings/stubs/binding_generation/src/main.rs [hunk 1]
```diff
@@ -35,143 +36,176 @@ use wires_15_stubs::{
 fn main() {
     let args: Vec<String> = std::env::args().collect();
 
-    if let Some(output_file) = args.get(1) {
-        let mut file = File::create(output_file).expect("could not create output file");
-        generate_bindings(&mut file);
+    let env = &mut Env::default();
+
+    let header = "(* This file is generated automatically with ocaml_gen. *)\n";
```
Comment (mimoo) [inline src/lib/crypto/kimchi_bindings/stubs/binding_generation/src/main.rs:41]: noice

File: src/lib/crypto/kimchi_bindings/stubs/binding_generation/src/main.rs [hunk 2]
```diff
@@ -35,143 +36,176 @@ use wires_15_stubs::{
 fn main() {
```
Comment (mimoo) [inline src/lib/crypto/kimchi_bindings/stubs/binding_generation/src/main.rs:36]: can you update the `README.md` in the stubs to document the new usage for this CLI?

File: src/lib/crypto/kimchi_bindings/stubs/binding_generation/src/main.rs [hunk 3]
```diff
@@ -35,143 +36,176 @@ use wires_15_stubs::{
...
+        decl_module!(w, env, "Kimchi_types", {
+            generate_types_bindings(&mut w, env);
+        });
     };
+    if let Some(pasta_bindings) = args.get(2) {
+        let mut file = File::create(pasta_bindings).expect("could not create output file");
+        write!(file, "{}", header).unwrap();
+        let _ = env.new_module("Pasta_bindings");
+        generate_pasta_bindings(&mut file, env);
+        let _ = env.parent();
```
Comment (mimoo) [inline src/lib/crypto/kimchi_bindings/stubs/binding_generation/src/main.rs:61]: Interesting, I was wondering how the bindings from Kimchi_bindings were going to refer to the bindings from Pasta_bindings. So I guess you fake the module here, I'm happy it works without any changes to ocaml-gen :)

File: src/lib/crypto/kimchi_bindings/pasta_fq_poseidon/kimchi_pasta_fq_poseidon.ml [hunk 4]
```diff
@@ -2,5 +2,5 @@ type t
 
 external create : unit -> t = "caml_pasta_fq_poseidon_params_create"
 
-external block_cipher : t -> Kimchi.FieldVectors.Fq.t -> unit
+external block_cipher : t -> Kimchi_bindings.FieldVectors.Fq.t -> unit
```
Comment (mimoo) [inline src/lib/crypto/kimchi_bindings/pasta_fq_poseidon/kimchi_pasta_fq_poseidon.ml:5]: it'd be nice to use ocaml-gen for both this pasta_fp_poseidon and pasta_fq_poseidon packages

**Review by mimoo — APPROVED** [review]
LGTM

**Comment by mrmr1993** [comment]
I've merged compatible and develop into here (by merging the branch from #10664) to try to front-run on CI. When that PR is merged, the extra changes here will disappear again.

**Pull request merged**